Lively/src/components/session/TeamEntryForm.tsx

```diff
@@ -0,0 +1,166 @@
+import { MutationFunc } from 'react-apollo';
+
+interface State {
+    email: string;
+    password: string;
+    loading: boolean;
+    badEmail: boolean; 
+    badPassword: boolean;
+    shouldFocus: boolean;
+}
+
+type MutationPayload = {
+    teamLogin: {
+        ok: boolean;
+        token?: string;
+        refreshToken?: string;
+    }
+};
+  
+type MutationInput = {
+    url: string;
+    password: string; 
+    email: string
+ };
+  
+interface Props {
+    url: string;
+    setError: (error: boolean) => void; 
+    teamLogin: MutationFunc<MutationPayload, MutationInput>;
+}
+
+class TeamEntryForm extends React.Component<Props, State> {
+    state = {
+        email: '', 
+        password: '', 
+        loading: false,
+        badEmail: false,
+        badPassword: false,
+        shouldFocus: false,
+    };
+
+    badInputStyle = {border: '1px solid #d72b3f', background: '#fbeaec'};
+
+    handleChange = (e: React.FormEvent<HTMLInputElement>) => {
+        this.setState({
+            [e.currentTarget.name as any]: e.currentTarget.value // tslint:disable-line
+        });
+    }
+
+    checkBadFields = (submitting = false) => {
+        const {email, password} = this.state;
+        let badInputFound = false;
+        if ((email === '' && password === '') && !submitting) {
+            return true;
+        } 
+        if (email === '' || !validateEmail(email)) {
+            this.setState({badEmail: true});
+            badInputFound = true;
+        } else {
+            this.setState({badEmail: false});
+        } 
+        if (password === '') {
+            this.setState({badPassword: true});
+            badInputFound = true;           
+        } else {
+            this.setState({badPassword: false});
+        }
+        return badInputFound;
+    }
+
+    handlePasswordFocus = () => {
+        // no red error if they type a correct email then click on password field
+        const { email, password, shouldFocus } = this.state;
+        if (validateEmail(email) && !shouldFocus) {
+            if (password === '') {
+                this.setState({badPassword: false});
+            }
+        }
+    }
+
+    focus = (type: string) => {
+        // if bad email or password is in state and form is submitted, focus on that input
+        let focusFunc = (input: HTMLInputElement) => input && input.focus();
+        let noFocus = (input: HTMLInputElement) => null;
+        if (!this.state.shouldFocus) { return noFocus; }
+        const {badEmail, badPassword} = this.state;
+        if (type === 'email') {
+            return badEmail ? focusFunc : noFocus;
+        } else if (!badEmail) {
+           return badPassword ? focusFunc : noFocus;
+        } else {
+            return noFocus;
+        }
+    }
+
+    handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
+        e.preventDefault();
+        this.setState({ shouldFocus: true });
+        setTimeout(() => this.setState({ shouldFocus: false }), 0);
```

Comment:
why use setTimeout? if this is 100% necessary, it deserves a comment explaining why